ext4/dmverity/dmverity.go

```diff
@@ -0,0 +1,179 @@
+	HashBlockSize uint32
+	DataBlocks    uint64
+}
+
+// MerkleTree constructs dm-verity hash-tree for a given byte array with a fixed salt (0-byte) and algorithm (sha256).
+func MerkleTree(data []byte) []byte {
+	layers := make([][]byte, 0)
+
+	var currentLevel = bytes.NewBuffer(data)
+
+	for currentLevel.Len() != blockSize {
+		var blocks = currentLevel.Len() / blockSize
+		var nextLevel = bytes.NewBuffer(make([]byte, 0))
+
+		for i := 0; i < blocks; i++ {
+			block := make([]byte, blockSize)
+			_, _ = currentLevel.Read(block)
+			h := hash2(salt, block)
+			nextLevel.Write(h)
+		}
+
+		padding := bytes.Repeat([]byte{0}, blockSize-(nextLevel.Len()%blockSize))
+		nextLevel.Write(padding)
+
+		currentLevel = nextLevel
+		layers = append(layers, currentLevel.Bytes())
+	}
+
+	var tree = bytes.NewBuffer(make([]byte, 0))
+	for i := len(layers) - 1; i >= 0; i-- {
+		tree.Write(layers[i])
+	}
+
+	return tree.Bytes()
+}
+
+// RootHash computes root hash of dm-verity hash-tree
+func RootHash(tree []byte) []byte {
+	return hash2(salt, tree[:blockSize])
+}
+
+// NewDMVeritySuperblock returns a dm-verity superblock for a device with a given size, salt, algorithm and versions are
+// fixed.
+func NewDMVeritySuperblock(size uint64) *dmveritySuperblock {
+	superblock := &dmveritySuperblock{
+		Version:       1,
+		HashType:      1,
+		UUID:          generateUUID(),
+		DataBlockSize: blockSize,
+		HashBlockSize: blockSize,
+		DataBlocks:    size / blockSize,
+		SaltSize:      uint16(len(salt)),
+	}
+
+	copy(superblock.Signature[:], "verity")
+	copy(superblock.Algorithm[:], "sha256")
+	copy(superblock.Salt[:], salt)
+
+	return superblock
+}
+
+func hash2(a, b []byte) []byte {
+	h := sha256.New()
+	h.Write(append(a, b...))
+	return h.Sum(nil)
+}
+
+func generateUUID() [16]byte {
+	res := [16]byte{}
+	if _, err := rand.Read(res[:]); err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// ReadDMVerityInfo extracts dm-verity super block information and merkle tree root hash
+func ReadDMVerityInfo(vhdPath string, offsetInBytes int64) (*VerityInfo, error) {
+	vhd, err := os.OpenFile(vhdPath, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer vhd.Close()
+
+	// Skip the ext4 data to get to dm-verity super block
+	if s, err := vhd.Seek(offsetInBytes, io.SeekStart); err != nil || s != offsetInBytes {
+		return nil, errors.Wrap(err, "failed to seek dm-verity super block")
+	}
+
+	block := make([]byte, blockSize)
+	if s, err := vhd.Read(block); err != nil || s != blockSize {
+		return nil, ErrSuperBlockReadFailure
+	}
+
+	var dmvSB = &dmveritySuperblock{}
+	b := bytes.NewBuffer(block)
+	if err := binary.Read(b, binary.LittleEndian, dmvSB); err != nil {
+		return nil, ErrSuperBlockParseFailure
+	}
+
+	// read the merkle tree root
+	if s, err := vhd.Read(block); err != nil || s != blockSize {
+		return nil, ErrRootHashReadFailure
+	}
+	rootHash := hash2(dmvSB.Salt[:dmvSB.SaltSize], block)
+	return &VerityInfo{
+		RootDigest:         fmt.Sprintf("%x", rootHash),
```

Comment:
instead of doing an sprintf, you can use golang's built in `string` func to convert a byte slice to a string. 

Ex:
```
buf := []byte{'a','b','c'} 
bufAsString := string(buf)
```

Reply:
This will have it try and be decoded as valid utf-8 I believe. Here's a run with the alpine image in the README for both methods.

`fmt.Sprintf("%x", rootHash)`: ea1303202ccc2e71b4f40d13537c0dda1058b232293d42c6cd18258df0210642 

Convert via string(): �►X�2)=B��↑%��!♠B

Reply:
It seems like this was intentional

Reply:
yeah, this was intentional. I initially did try to just convert via `string`.